File: sdk/samples/WFSamples/Technologies/Markup/CustomSerialization/VB/CollectionActivities/QueueSerializer.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------


Imports System
Imports System.Collections
Imports System.ComponentModel.Design.Serialization
Imports System.Runtime.InteropServices
Imports System.Workflow.ComponentModel.Serialization

<ComVisible(False)> _
Public Class QueueSerializer
    Inherits WorkflowMarkupSerializer
    Implements IDesignerSerializationProvider

    Protected Overrides Sub AddChild(ByVal serializationManager As WorkflowMarkupSerializationManager, ByVal parentObject As Object, ByVal childObject As Object)

        If parentObject Is Nothing Then
            Throw New ArgumentNullException("parentObject")
        End If

        If (childObject Is Nothing) Then
            Throw New ArgumentNullException("childObject")
        End If

        Dim queue As Queue = TryCast(parentObject, Queue)

        If queue Is Nothing Then
            Throw New Exception(String.Format("The type of parentObject is not {0}", GetType(Queue).FullName))
        End If

        queue.Enqueue(childObject)
    End Sub

    Protected Overrides Function GetChildren(ByVal serializationManager As WorkflowMarkupSerializationManager, ByVal obj As Object) As IList

        If obj Is Nothing Then
            Throw New ArgumentNullException("obj")
        End If

        Dim queue As Queue = TryCast(obj, Queue)

        If queue Is Nothing Then
            Throw New Exception(String.Format("The type of obj is not {0}", GetType(Queue).FullName))
        End If

        Dim arrayList As New ArrayList(queue.ToArray())

        Return arrayList
    End Function

#Region "IDesignerSerializationProvider Members"
    Public Function GetSerializer(ByVal manager As System.ComponentModel.Design.Serialization.IDesignerSerializationManager, ByVal currentSerializer As Object, ByVal objectType As System.Type, ByVal serializerType As System.Type) As Object Implements System.ComponentModel.Design.Serialization.IDesignerSerializationProvider.GetSerializer
        If objectType Is GetType(Queue) Then
            Return Me
        Else
            Return currentSerializer
        End If
    End Function
#End Region

End Class

File: sdk/samples/WFSamples/Technologies/Markup/CustomSerialization/CS/CollectionActivities/QueueSerializer.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.Collections;
using System.ComponentModel.Design.Serialization;
using System.Workflow.ComponentModel.Serialization;

namespace Microsoft.Samples.Workflow.CustomSerialization
{
    public class QueueSerializer : WorkflowMarkupSerializer, IDesignerSerializationProvider
    {
        protected override void AddChild(WorkflowMarkupSerializationManager serializationManager, object parentObject, object childObject)
        {
            if (parentObject == null)
                throw new ArgumentNullException("parentObject");

            if (childObject == null)
                throw new ArgumentNullException("childObject");

            Queue queue = parentObject as Queue;
            if (queue == null)
                throw new Exception(string.Format("The type of parentObject is not {0}", typeof(Queue).FullName));

            queue.Enqueue(childObject);
        }

        protected override IList GetChildren(WorkflowMarkupSerializationManager serializationManager, object obj)
        {
            if (obj == null)
                throw new ArgumentNullException("obj");

            Queue queue = obj as Queue;
            if (queue == null)
                throw new Exception(string.Format("The type of obj is not {0}", typeof(Queue).FullName));

            ArrayList arrayList = new ArrayList(queue.ToArray());
            return arrayList;
        }

        #region IDesignerSerializationProvider Members

        public object GetSerializer(IDesignerSerializationManager manager, object currentSerializer, Type objectType, Type serializerType)
        {
            if (objectType == typeof(Queue))
                return this;
            else
                return currentSerializer;
        }

        #endregion
    }
}
